### PR TITLE
Fix flaky unit tests

### LINKED ISSUE #3624: RestoreRetryTests tests are flaky
RestoreRetryTests tests are flaky.  Example failures from test runs:

```
RestoreRetryTests.RestoreRetry_PackagesConfigRetryOnFailingV3Source     

Restoring NuGet package testPackage1.1.1.0.
WARNING: Unable to find version '1.1.0' of package 'testPackage1'.
  http://localhost:50231/index.json: Value cannot be null.
Parameter name: baseUrl


NuGet Config files used:
    C:\BuildAgent\temp\buildTmp\NuGetTestFolder\80bf1c2e-7079-4fa2-be3c-10aff91e893a\NuGet.Config
    C:\Users\fxsign\AppData\Roaming\NuGet\NuGet.Config
    C:\ProgramData\nuget\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    http://localhost:50231/index.json

Errors in packages.config projects
    Unable to find version '1.1.0' of package 'testPackage1'.
      http://localhost:50231/index.json: Value cannot be null.
    Parameter name: baseUrl

Expected: True
Actual:   False
   at NuGet.CommandLine.Test.RestoreRetryTests.RestoreRetry_PackagesConfigRetryOnFailingV3Source() in C:\BuildAgent\work\67659811269ba07e\NuGet.Client\test\NuGet.Clients.Tests\NuGet.CommandLine.Test\RestoreRetryTests.cs:line 580
```

```
 RestoreRetryTests.RestoreRetry_ProjectJsonRetryOnFailingV3Source   

SBuild auto-detection: using msbuild version '15.0' from 'C:\Program Files (x86)\MSBuild\15.0\bin\amd64'.
Restoring packages for C:\BuildAgent\temp\buildTmp\NuGetTestFolder\9f398fea-267d-4b89-aae1-7cf17a4cb6ff\project.json...
 Value cannot be null.
Parameter name: baseUrl

Expected: True
Actual:   False
   at NuGet.CommandLine.Test.RestoreRetryTests.RestoreRetry_ProjectJsonRetryOnFailingV3Source() in C:\BuildAgent\work\67659811269ba07e\NuGet.Client\test\NuGet.Clients.Tests\NuGet.CommandLine.Test\RestoreRetryTests.cs:line 385
```

### PR DESCRIPTION
**Pull request by dtivel**
Fix NuGet/Home#3624.

The issue is that concurrent tests sometimes interfere with each other.  This fix has two parts:
1.  Clear all caches before starting tests and again on successful test completion.  If there are test failures, caches are left as-is for post-mortem investigation.
2.  Ensure each URL used by `MockServer` is unique.  This change is transparent to most tests.  However, for tests which access the requested URL, some string pruning is necessary to strip off the random base path.

Also, reenable test `RestoreRetry_PackagesConfigRetryOnFailingV2Source` as it appears to pass consistently.

@rrelyea @emgarten @joelverhagen @drewgil @alpaix @mishra14 @jainaashish @zhili1208 @rohit21agrawal

### CODE REVIEW

File: runTests.ps1 [hunk 1]
```diff
@@ -85,6 +85,11 @@ if (-not $SkipVS15 -and -not $VS15Installed) {
 
 $BuildErrors = @()
 
+Invoke-BuildStep 'Clear local caches' {
+        & nuget locals all -clear -verbosity detailed
```
Comment (alpaix) [inline runTests.ps1]: `Clear-PackageCache` in `build/common.ps1` does the same (almost).

File: runTests.ps1 [hunk 2]
```diff
@@ -147,6 +152,10 @@ Trace-Log ('=' * 60)
 if ($BuildErrors) {
     $ErrorLines = $BuildErrors | %{ ">>> $($_.Exception.Message)" }
     Write-Error "Build's completed with $($BuildErrors.Count) error(s):`r`n$($ErrorLines -join "`r`n")" -ErrorAction Stop
+} else {
```
Comment (alpaix) [inline runTests.ps1]: `else` is not needed as `Write-Error ... -ErrorAction Stop` would terminate execution of the script.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs [hunk 3]
```diff
@@ -401,16 +407,18 @@ public void Dispose()
     /// </remarks>
     public class RouteTable
     {
-        List<Tuple<string, Func<HttpListenerRequest, object>>> _mappings;
+        private string _basePath;
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs]: `readonly`

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs [hunk 4]
```diff
@@ -401,16 +407,18 @@ public void Dispose()
...
+        private List<Tuple<string, Func<HttpListenerRequest, object>>> _mappings;
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/MockServer.cs]: `readonly`

File: runTests.ps1 (hunk 1 above)
Comment (dtivel) [reply runTests.ps1]: `Clear-PackageCache` does not clear the `http-cache`.  Why is that?

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetListCommandTest.cs [hunk 5]
```diff
@@ -639,7 +639,7 @@ public void ListCommand_SimpleV3()
                 {
                     serverV3.Get.Add("/", r =>
                     {
-                        var path = r.Url.AbsolutePath;
+                        var path = r.Url.AbsolutePath.Substring(serverV3.BasePath.Length);
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetListCommandTest.cs]: This statement repeats itself many times and it's hard to memorize.
Consider adding a helper method like `MockServer.GetPath(HttpRequest r);`

Comment (alpaix) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetListCommandTest.cs]: Plus, if it'd be added it'd be great adding some explanation to the new method what it does and why it's needed.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs [hunk 6]
```diff
@@ -28,14 +28,16 @@ public class PortReserver : IDisposable
...
-        }
+        public int PortNumber { get; private set; }
+        public string BaseUri { get; }
 
-        public PortReserver(int basePort = 50231)
+        public PortReserver(string basePath = null, int basePort = 50231)
         {
+            if (!string.IsNullOrEmpty(basePath) && (!basePath.StartsWith("/") || basePath.EndsWith("/")))
+            {
+                throw new ArgumentException($"If provided, argument \"{nameof(basePath)}\" must start with and must not end with a slash (/)");
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/PortReserver.cs]: Consider adding documentation to the constructor explaining correct usage of the parameters.

File: runTests.ps1 (hunk 1 above)
Comment (alpaix) [reply runTests.ps1]: Consider changing it to do whatever is needed.

**Review by alpaix — APPROVED** [review]
LGTM

**Pull request merged**